src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java

```diff
@@ -0,0 +1,522 @@
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.xpath;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtils;
+import net.sf.saxon.Configuration;
+import net.sf.saxon.event.Receiver;
+import net.sf.saxon.expr.parser.Location;
+import net.sf.saxon.om.AtomicSequence;
+import net.sf.saxon.om.Item;
+import net.sf.saxon.om.NamespaceBinding;
+import net.sf.saxon.om.NodeInfo;
+import net.sf.saxon.om.SequenceIterator;
+import net.sf.saxon.om.TreeInfo;
+import net.sf.saxon.pattern.NodeTest;
+import net.sf.saxon.trans.XPathException;
+import net.sf.saxon.tree.iter.AxisIterator;
+import net.sf.saxon.tree.util.FastStringBuffer;
+import net.sf.saxon.tree.util.Navigator;
+import net.sf.saxon.type.SchemaType;
+
+/**
+ * Represents general class for {@code ElementNode}, {@code RootNode} and {@code AttributeNode}.
+ *
+ * @author Timur Tibeyev
+ */
+// -@cs[MethodCount] This class implements all interface methods. Total number of methods is 47.
+// No way to decrease total number of methods, because interface is the part of the Saxon library.
+public abstract class AbstractNode implements NodeInfo {
+    /** The map of {@link TokenTypes} to {@code String}. */
+    private static final Map<Integer, String> TOKEN_TYPES_TO_STRING_MAP = new HashMap<>();
+
+    /** The children. */
+    private final List<AbstractNode> children = new ArrayList<>();
+
+    /** The token type. */
+    private int tokenType;
+
+    /** The line number. */
+    private int lineNo;
+
+    /** The column number. */
+    private int columnNo;
+
+    /**
+     * Getter method for token type.
+     * @return token type
+     */
+    public int getTokenType() {
+        return tokenType;
+    }
+
+    /**
+     * Setter method for token type.
+     * @param tokenType token type
+     */
+    protected void setTokenType(int tokenType) {
```

Comment:
@romani 
Yes, should be made immutable.

@timurt 
Please, do for all similar cases. Remove as much setXXX methods as possible. Try to make all nodes immutable and fully initialied them via constructor. So fields should become final if possible.